File: Problems/TwoPointersType/Valid_Palindrome/ValidPalindrome.java
package TwoPointersType.Valid_Palindrome;

public class ValidPalindrome {
    public static void main(String[] args) {
        String[] test = new String[]{"A man, a plan, a canal: Panama", "race a car", " ", "0P"};
        Solution sol = new Solution();
        boolean[] expect = new boolean[]{true, false, true, false};

        System.out.println("=== Valid Palindrome Test Results ===");

        int passed = 0;

        for (int i = 0; i < test.length; i++) {
            boolean got = sol.isPalindrome(test[i]);
            boolean ok = (got == expect[i]);

            System.out.printf(
                "[Case %d]%nInput: \"%s\"%nExpected: %s%nGot: %s%nResult: %s%n%n",
                i + 1, test[i], expect[i], got, ok ? "PASS ✅" : "FAIL ❌"
            );

            if (ok) passed++;
        }

        System.out.printf("Summary: %d/%d test cases passed.%n", passed, test.length);

    }
}

class Solution {
    public boolean isPalindrome(String s) {
        if (s.length() <= 1) return true;

        // two-pointer thought
        int frontIdx = 0;
        int endIdx = s.length() - 1;

        while (frontIdx < endIdx){
            char front = Character.toUpperCase(s.charAt(frontIdx));
            char end = Character.toUpperCase(s.charAt(endIdx));

            // check are both idx are letter
            if (!Character.isLetterOrDigit(front)){
                frontIdx++;
                continue;
            } else if (!Character.isLetterOrDigit(end)) {
                endIdx--;
                continue;
            }

            if (front != end) return false;
            frontIdx++;
            endIdx--;
        }

        return true;
    }
}

File: Problems/TwoPointersType/Valid_Palindrome/valid_palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        front_idx:int = 0
        end_idx:int = len(s) - 1

        while (front_idx < end_idx):
            front:str = s[front_idx]
            end:str = s[end_idx]

            if not front.isalnum():
                front_idx += 1
                continue
            elif not end.isalnum():
                end_idx -= 1
                continue
            
            if front.upper() != end.upper(): return False

            front_idx += 1
            end_idx -= 1

        return True

if __name__ == "__main__":
    test: list[str] = ["A man, a plan, a canal: Panama", "race a car", " ", "0P"]
    expect: list[bool] = [True, False, True, False]
    passed_count: int = 0
    
    for i in range(len(test)):
        res_flag:bool = Solution().isPalindrome(test[i])
        varify:bool = res_flag == expect[i]
        
        # python way ternary conditional expression
        pass_state:str = "PASS ✅" if varify else "FAIL ❌"
        
        passed_count += 1 if varify == True else 0
        print(f"[Case {i + 1}]\nInput: {test[i]}\nExpected: {expect[i]}\nGot: {varify}\nResult: {pass_state}\n")
    
    print(f"Summary: {passed_count}/{len(test)} test cases passed.")